sql/parser/builtins.go

```diff
@@ -62,103 +62,92 @@ type builtin struct {
 // Need to figure out if there is a standard set of functions or if we can
 // customize this to our liking. Would be good to support type conversion
 // functions.
-var builtins = map[string]builtin{
+var builtins = map[string][]builtin{
 	// TODO(XisiHuang): support encoding, i.e., length(str, encoding).
-	"length": stringBuiltin(func(s string) (Datum, error) {
+	"length": {stringBuiltin(func(s string) (Datum, error) {
 		return DInt(len(s)), nil
-	}),
+	})},
 
-	"lower": stringBuiltin(func(s string) (Datum, error) {
+	"lower": {stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToLower(s)), nil
-	}),
+	})},
 
-	"upper": stringBuiltin(func(s string) (Datum, error) {
+	"upper": {stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
-	}),
+	})},
 
 	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax.
 	"substring": {
-		nArgs: -1,
-		fn: func(args DTuple) (Datum, error) {
-			argsNum := len(args)
-			if argsNum != 2 && argsNum != 3 {
-				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
-					argsNum, "2 or 3")
-			}
-
-			dstr, ok := args[0].(DString)
-			if !ok {
-				return DNull, argTypeError(args[0], dstr.Type())
-			}
-			str := string(dstr)
-
-			start, ok := args[1].(DInt)
-			if !ok {
-				return DNull, argTypeError(args[1], start.Type())
-			}
-			s := int(start)
-
-			var e int
-			if argsNum == 2 {
-				e = len(str) + 1
-			} else {
-				count, ok := args[2].(DInt)
-				if !ok {
-					return DNull, argTypeError(args[2], count.Type())
+		builtin{
+			types: []string{"string", "int"},
```

Comment:
Might be better/safer/faster to use `reflect.Type` instead of strings here. We already have `boolType`, `intType`, etc globals (see `eval.go`). This would become:

```
  types: []reflect.Type{stringType, intType}
```

This more explicitly reflects are desires and the `reflect.Type` equality comparisons are simple pointer comparisons that will be faster than the string comparisons.